application/src/Api/Adapter/AbstractEntityAdapter.php

```diff
@@ -272,13 +272,21 @@ public function search(Request $request)
 
         // Add the LIMIT clause.
         $this->limitQuery($qb, $query);
+        $maxResults = $qb->getMaxResults();
+        $firstResult = $qb->getFirstResult();
 
         // Before adding the ORDER BY clause, set a paginator responsible for
-        // getting the total count. This optimization excludes the ORDER BY
-        // clause from the count query, greatly speeding up response time.
-        $countQb = clone $qb;
-        $countQb->select('1')->resetDQLPart('orderBy');
-        $countPaginator = new Paginator($countQb, false);
+        // getting the total count (unless configured not to or when no pagination
+        // is set.
+        $countQueryDefault = $maxResults !== null || $firstResult > 0;
+        $countQuery = $request->getOption('countQuery') ?? $countQueryDefault;
```

Comment:
getOption takes as its second arg the default value if I'm remembering correctly... should be the same behavior as this using the `??` but I think we'd want to keep to that established pattern